FreeRTOS-Plus/Demo/coreMQTT_Windows_Simulator/MQTT_Multitask/DemoTasks/simple_sub_pub_demo.c

```diff
@@ -0,0 +1,546 @@
+    {
+        ulNextSubscribeMessageID++;
+        ulSubscribeMessageID = ulNextSubscribeMessageID;
+    }
+    taskEXIT_CRITICAL();
+
+    /* Complete the subscribe information.  The topic string must persist for
+     * duration of subscription! */
+    xSubscribeInfo.pTopicFilter = pcTopicFilter;
+    xSubscribeInfo.topicFilterLength = ( uint16_t ) strlen( pcTopicFilter );
+    xSubscribeInfo.qos = xQoS;
+    xSubscribeArgs.pSubscribeInfo = &xSubscribeInfo;
+    xSubscribeArgs.numSubscriptions = 1;
+
+    /* Complete an application defined context associated with this subscribe message.
+     * This gets updated in the callback function so the variable must persist until
+     * the callback executes. */
+    xApplicationDefinedContext.ulNotificationValue = ulNextSubscribeMessageID;
+    xApplicationDefinedContext.xTaskToNotify = xTaskGetCurrentTaskHandle();
+    xApplicationDefinedContext.pArgs = ( void * ) &xSubscribeArgs;
+
+    xCommandParams.blockTimeMs = mqttexampleMAX_COMMAND_SEND_BLOCK_TIME_MS;
+    xCommandParams.cmdCompleteCallback = prvSubscribeCommandCallback;
+    xCommandParams.pCmdCompleteCallbackContext = ( void * ) &xApplicationDefinedContext;
+
+    /* Loop in case the queue used to communicate with the MQTT agent is full and
+     * attempts to post to it time out.  The queue will not become full if the
+     * priority of the MQTT agent task is higher than the priority of the task
+     * calling this function. */
+    LogInfo( ( "Sending subscribe request to agent for topic filter: %s with id %d",
+               pcTopicFilter,
+               ( int ) ulSubscribeMessageID ) );
+
+    do
+    {
+        /* TODO: prvIncomingPublish as publish callback. */
+        xCommandAdded = MQTTAgent_Subscribe( &xGlobalMqttAgentContext,
+                                             &xSubscribeArgs,
+                                             &xCommandParams );
+    } while( xCommandAdded != MQTTSuccess );
+
+    /* Wait for acks to the subscribe message - this is optional but done here
+     * so the code below can check the notification sent by the callback matches
+     * the ulNextSubscribeMessageID value set in the context above. */
+    xCommandAcknowledged = prvWaitForCommandAcknowledgment( NULL );
+
+    /* Check both ways the status was passed back just for demonstration
+     * purposes. */
+    if( ( xCommandAcknowledged != pdTRUE ) ||
+        ( xApplicationDefinedContext.xReturnStatus != MQTTSuccess ) )
+    {
+        LogInfo( ( "Error or timed out waiting for ack to subscribe message topic %s",
+                   pcTopicFilter ) );
+    }
+    else
+    {
+        LogInfo( ( "Received subscribe ack for topic %s containing ID %d",
+                   pcTopicFilter,
+                   ( int ) xApplicationDefinedContext.ulNotificationValue ) );
+    }
+
+    return xCommandAcknowledged;
+}
+
+/*-----------------------------------------------------------*/
+volatile uint32_t ulQoS0FailureCount = 0UL;
```

Comment:
Should this variable be defined at the top of the file along with documentation?

Reply:
I wanted to keep this in sync with the demo in labs, but I'll just remove Richard's changes if they're going to cause this many issues